pkg/registry/discovery/endpointslice/strategy.go

```diff
@@ -186,11 +188,40 @@ func dropTopologyOnV1(ctx context.Context, oldEPS, newEPS *discovery.EndpointSli
 			return
 		}
 
+		// Only node names that exist in previous version of the EndpointSlice
+		// deprecatedTopology fields may be retained in new version of the
+		// EndpointSlice.
+		prevNodeNames := getDeprecatedTopologyNodeNames(oldEPS)
+
 		for i := range newEPS.Endpoints {
 			ep := &newEPS.Endpoints[i]
 
-			//Silently clear out DeprecatedTopology
-			ep.DeprecatedTopology = nil
+			newTopologyNodeName, ok := ep.DeprecatedTopology[corev1.LabelHostname]
+			if ep.NodeName == nil && ok && prevNodeNames.Has(newTopologyNodeName) {
+				// Retain node name in deprecatedTopology until the node name field has been populated
+				ep.DeprecatedTopology = map[string]string{corev1.LabelHostname: newTopologyNodeName}
+			} else {
+				// Silently clear out DeprecatedTopology
+				ep.DeprecatedTopology = nil
+			}
+		}
+	}
+}
+
+// getDeprecatedTopologyNodeNames returns a set of node names present in
+// deprecatedTopology fields within the provided EndpointSlice.
+func getDeprecatedTopologyNodeNames(eps *discovery.EndpointSlice) sets.String {
+	if eps == nil {
+		return nil
+	}
+	var names sets.String
+	for _, ep := range eps.Endpoints {
+		if nodeName, ok := ep.DeprecatedTopology[corev1.LabelHostname]; ok && len(nodeName) > 0 {
+			if names == nil {
+				names = sets.NewString()
```

Comment:
premature optimization?

Reply:
I suggested this since 1) we expect the number of old endpoint slices with these label maps to go to zero, 2) endpointslices are a high-frequency write target, and 3) map allocations are more expensive than typical structs

Reply:
ok